sat-3016  - Added/Edited new chapter/assembly about foreman webhooks

neal-timpe opened the pull request:
Edited previous work on foreman webhooks.

[doc] Foreman Webhooks

https://issues.redhat.com/browse/SAT-3106

Cherry-pick into:

* [x] Foreman 2.5 (Satellite 6.10)
* [ ] Foreman 2.4
* [ ] Foreman 2.3 (Satellite 6.9)
* [ ] Foreman 2.1 (Satellite 6.8)

<!---
Thank you for contributing to Foreman documentation. Make sure to read README
for the documentation standards. Set cherry-pick github label to mark this
contribution for cherry picking and check which version do you need with [x].
-->

melcorr commented:
Hello and welcome.

I am... confused... - we have also PR #612 

What is occurring :o)

neal-timpe commented:
Hi @melcorr , @ianf77 and the team asked me to work on this so Ian could work on other things. I've recreated his PR and attempted to address some of the concerns that you had raised in #612 .

melcorr commented:
> Hi @melcorr , @ianf77 and the team asked me to work on this so Ian could work on other things. I've recreated his PR and attempted to address some of the concerns that you had raised in #612 .

Hey @neal-timpe thank you for clarifying. 
Are you able to build the docs locally and test that the attributes are correct and render and the URLs etc are OK? That would be very helpful. I can already see some feedback from the previous PR that was not applied. 

If you haven't set yourself up for local tests, please follow: https://github.com/theforeman/foreman-documentation/blob/master/guides/README.md#building 

I appreciate you helping out!

neal-timpe commented:
@melcorr Yes, I set up the tools and ran a build. It looks okay to me and the links work.

I also went back through to make sure there was one sentence per line. 
I added some context to the creating a webhook module
I also removed the templates module after chatting with Ian. 

What else do you see? Which items did I miss?

melcorr commented on guides/common/modules/proc_creating-a-webhook.adoc:
```diff
@@ -0,0 +1,55 @@
+. Click *Subscribe to field* to select an event.
+. Enter a name.
+. Enter a target URL.
+  When using the shellhooks plugin, the URL should be in the form of `https://{smartproxy.example.com}:{smartproxy_port}/shellhooks/my_script`.
```
```suggestion
  When using the shellhooks plugin, the URL should be in the form of `https://{smartproxy-example-com}:{smartproxy_port}/shellhooks/my_script`.
```
This wasn't implemented from the previous review.

neal-timpe replied:
Ok, I see that now. I fixed that in two places. Thanks!

melcorr reviewed (commented):
Hey @neal-timpe - thanks so much. 

I pointed to the attribute that I immediately saw and had pointed out in the other PR. There is no way this can build correctly. 

Now you have confirmed you have built locally, I will review this in the morning. 

I appreciate you getting back to me and your contribution. 
Thanks, 
Melanie

melcorr commented on guides/common/modules/ref_shellhooks.adoc:
```diff
@@ -0,0 +1,16 @@
+[id="shellhooks_{context}"]
+= Shellhooks
+
+You can install a {SmartProxy} Shellhooks plugin that exposes executables using a REST HTTPS API.
```
```suggestion
You can install a {SmartProxy} shellhooks that exposes executables using a REST HTTPS API.
```

melcorr reviewed (changes requested):
Hey Neal, 
Firstly, this is a great first contribution and thank you for being responsible to my earlier comments, for confirming that you have built locally, and for implementing the initial feedback. It's not easy to come into a project half way through a work in progress. I appreciate your efforts and responsiveness. 

Overall, we are very much on track with this. 

My major concern is that the information about webhooks and shellhooks should be treated separately and have separate procedures. 
The engineer's notes probably had them jumbled together, but this isn't really standard throughout our docset. You could either clone the files (easiest) and create shellhooks procedures, or conditionalise the files and have them print twice depending on the context set. As it is now, in the organization of the chapter. This really doesn't seem consistent with the rest of our docs. However, it's an easy enough thing to fix! 

If you're unsure about anything, let me know. 

@ofedoren - in the absence of lzap, would you mind having a quick look through these procedures? If you can, could you confirm how `foreman_hooks` is referred to in the downstream? Can you give a general + or - to these docs?

ofedoren commented on guides/common/modules/proc_creating-a-webhook.adoc:
```diff
@@ -0,0 +1,55 @@
+Keep this field blank when using shellhooks.
+. Click the *Additional* tab.
+. Enter the HTTP Content Type.
+  For example, application/json, application/xml or text/plain depending on the payload.
```
I'd add `... on the payload you define. The application will not try to convert the content to match the specified content type.`

ofedoren commented on guides/common/modules/ref_shellhooks.adoc:
```diff
@@ -0,0 +1,16 @@
+
+The HTTP method must be POST. An example URL would be: `https://{smartproxy-example-com}:{smartproxy_port}/shellhooks/my_script`.
+
+Authorization with {Project} client checkbox must be enabled for each webhook to authorize a call.
```
Probably it's a mention of `Proxy Authorization` checkbox in the webhook form:
![ScreenShot-1627553978524](https://user-images.githubusercontent.com/32508194/127475407-3ddbf1e9-2376-4e55-a019-ecfdceeff5dc.png)

As help tooltip says: `Authorize with Foreman client certificate and validate smart-proxy CA from Settings`. AFAIK it needs to be enabled to make calls (send webhooks) to the smart proxy with shellhooks feature.

melcorr replied:
Thanks so much! To be in line with IBM style, please change this sentence to: 

```suggestion
Select the *Proxy Authorization* check box to enable each webhook to authorize a call.
```

ofedoren commented on guides/common/modules/proc_installing-webhooks-shellhooks-plugin.adoc:
```diff
@@ -0,0 +1,28 @@
+= Installing the Webhooks Plugin
+
+Use the following procedure to install the webhooks plugin. 
+After installing the `foreman_webhooks` plugin, you can configure {ProjectServer} to send webhook requests.
```
Ehm, not sure what you mean exactly. From the user perspective it will be called just `Webhooks`, but in installer (and other places) it will follow the same pattern as we do for every plugin, e.g. to install one will use `--enable-foreman-plugin-webhooks`.

melcorr replied:
I was just wondering would they be called Satellite hooks in the downstream ;-) 
We can just refer to them as webhooks, to be safe.

ofedoren commented on guides/common/modules/proc_creating-a-webhook-template.adoc:
```diff
@@ -0,0 +1,19 @@
+
+.Procedure
+
+. In the {ProjectWebUI}, navigate to *Administer* -> *Webhooks Templates*.
```
nit: change `->` to `>`

ofedoren commented on guides/common/modules/proc_creating-a-webhook.adoc:
```diff
@@ -0,0 +1,55 @@
+
+. In the {ProjectWebUI}, click *Administer* -> *Webhooks*.
+. Click *Create Webhook*.
+. Click *Subscribe to field* to select an event.
```
`field` should not be bold, the field is called just `Subscribe to`.

ofedoren commented on guides/common/modules/proc_creating-a-webhook.adoc:
```diff
@@ -0,0 +1,55 @@
+. Click the *Credentials* tab.
+. Optional: Enter the username and password when HTTP authentication is required.
++
+[NOTE]
```
I agree, but it's worth to have few notes to mention in the documentation, probably those things should be mentioned in the shellhooks configuration section?

Webhooks are designed to be more or less generic and can be used without shellhooks, but shellhooks plugin was created as part of foreman_hooks substitution.

ofedoren commented on guides/common/modules/proc_creating-a-webhook.adoc:
```diff
@@ -0,0 +1,55 @@
+. Click *Create Webhook*.
+. Click *Subscribe to field* to select an event.
+. Enter a name.
+. Enter a target URL.
```
Not sure how to mention this exactly, but `Target URL` supports simple ERB in it. For example you can configure a "dynamic" URL:

`Target URL - https://server.example.com/api/hosts/<%= @object.id %>/play_roles`

ofedoren commented on guides/common/modules/proc_installing-webhooks-shellhooks-plugin.adoc:
```diff
@@ -0,0 +1,28 @@
++
+[options="nowrap" subs="+quotes,attributes"]
+----
+# yum install tfm-rubygem-hammer_cli_foreman_webhooks
```
Currently it's the only way, but there is a set of patches to allow users to install the hammer plugin via installer as well.

ofedoren commented on guides/common/modules/ref_webhooks-available-events.adoc:
```diff
@@ -0,0 +1,45 @@
+[cols="50%,50%",options="header"]
+|====
+|Event name |Description
+|Actions Katello Content View Promote Succeeded |A content view was successfully promoted.
+|Actions Katello Content View Publish Succeeded |A repository was successfully synchronized.
+|Actions Remote Execution Run Host Job Succeeded |A generic remote execution job succeeded for a host. This event is emitted for all Remote Execution jobs, when complete
+|Actions Remote Execution Run Host Job Katello Errata Install Succeeded |Install errata using the Katello interface
+|Actions Remote Execution Run Host Job Katello Group Install Succeeded |Install package group using the Katello interface
+|Actions Remote Execution Run Host Job Katello Package Install Succeeded |Install package using the Katello interface
+|Actions Remote Execution Run Host Job Katello Group Remove |Remove package group using the Katello interface
+|Actions Remote Execution Run Host Job Katello Package Remove Succeeded |Remove package using the Katello interface
+|Actions Remote Execution Run Host Job Katello Service Restart Succeeded |Restart Services using the Katello interface
+|Actions Remote Execution Run Host Job Katello Group Update Succeeded |Update package group using the Katello interface
+|Actions Remote Execution Run Host Job Katello Package Update Succeeded |Update package using the Katello interface
+|Actions Remote Execution Run Host Job Foreman OpenSCAP Run Scans Succeeded |Run OpenSCAP scan
+|Actions Remote Execution Run Host Job Ansible Run Host Succeeded |Runs an Ansible playbook containing all the roles defined for a host
+|Actions Remote Execution Run Host Job Ansible Run Capsule Upgrade Succeeded |Upgrade Capsules on given Capsule server hosts
+|Actions Remote Execution Run Host Job Ansible Configure Cloud Connector Succeeded |Configure Cloud Connector on given hosts
+ifdef::satellite[]
+|Actions Remote Execution Run Host Job Ansible Run Insights Plan Succeeded |Runs a given maintenance plan from Red Hat Access Insights given an ID.
+endif::[]
+|Actions Remote Execution Run Host Job Ansible Run Playbook Succeeded |Run an Ansible playbook against given hosts
+|Actions Remote Execution Run Host Job Ansible Enable Web Console Succeeded |Run an Ansible playbook to enable the web console on given hosts
+|Actions Remote Execution Run Host Job Puppet Run Host Succeeded |Perform a single Puppet run
+|Actions Remote Execution Run Host Job Katello Module Stream Action Succeeded |Perform a module stream action using the Katello interface
+|Actions Remote Execution Run Host Job Leapp Pre-upgrade Succeeded |Upgradeability check for RHEL 7 host
+|Actions Remote Execution Run Host Job Leapp Remediation Plan Succeeded |Run Remediation plan with Leapp
+|Actions Remote Execution Run Host Job Leapp Upgrade Succeeded |Run Leapp upgrade job for RHEL 7 host
```
All those `Action` events will trigger webhooks only on `success`, so if an action failed, a webhook is not triggered.

ofedoren reviewed (commented):
Few more nitpicks.

neal-timpe commented on guides/common/modules/proc_installing-webhooks-shellhooks-plugin.adoc:
```diff
@@ -0,0 +1,28 @@
+# {foreman-installer} --enable-foreman-plugin-webhooks
+----
+
+* Optionally, install the shellhooks plugin on each {SmartProxy} used for shellhooks, using the following command:
```
I agree. What if we added a sentence that said, for more information, see the shellhooks and have an xref to the shellhooks procedure.

neal-timpe commented:
@melcorr I've addressed your comments. I'm not sure if I've chosen the correct title for the Shellhooks reference topic. I chose Shellhooks usage details. I acknowledge that may not be it, but I wanted to move forward. Anyway. I think I'm ready for another look. I also responded to one of your comments that is unresolved. I wanted to ask if that's an acceptable solution before I did it. Thanks!

melcorr commented:
@neal-timpe I will review this in the morning and get back to you!

melcorr commented on guides/common/assembly_using-foreman-webhooks.adoc:
```diff
@@ -0,0 +1,39 @@
+For example, when a user places an online order, a webhook makes an HTTP request to a pre-configured URL. 
+The request usually contains details of the event.
+
+An event triggers callbacks, such as sending an e-mail confirming an online purchase. 
```
My feedback wasn't applied here. Please let me know if there is something you don't understand.

melcorr replied:
OK, @neal-timpe - can you build locally? 
I just pulled your branch and am previewing locally but don't see these changes yet see them in the commits. 
Can you confirm?

melcorr reviewed (commented):
Can you please use this standard when writing commit messages: 

https://chris.beams.io/posts/git-commit/

melcorr reviewed (commented):
Neal can you take a look at your local setup please? 
I've pulled your fork, am in sync with your commits, but when I preview locally, I don't see the changes. I've rechecked a few times now, make clean and rebuilding.

melcorr commented:
Just for the record, I cloned your fork to a fresh folder, downloaded, checked your logs, and still this does not look right.

neal-timpe commented:
@melcorr I'm not sure what to check here. What doesn't look right about it? 

I have built locally and it looks fine to me. What should I be looking for?

melcorr commented on guides/common/assembly_using-foreman-webhooks.adoc:
```diff
@@ -0,0 +1,37 @@
+
+A webhook is a way for a web page or web application to provide other applications with information in real time. 
+Webhooks are only triggered after an event occurs. 
+For example, when a user places an online order, a webhook makes an HTTP request to a pre-configured URL. 
```
Can you remove this example as you have explained it in a Foreman context just below and this is not needed. It would be great to incorporate the  `a webhook makes an HTTP request to a pre-configured URL.` into the example below.

melcorr commented on guides/common/modules/proc_creating-a-webhook.adoc:
```diff
@@ -0,0 +1,55 @@
++
+. Select *Proxy Authorization* when using shellhooks, otherwise deselect this box.
+. Optional: Provide one or more CA certificates in the X509 Certification Authorities.
+Keep this field blank when using shellhooks.
```
This is still here. Please delete.

melcorr commented on guides/common/modules/proc_creating-a-webhook.adoc:
```diff
@@ -0,0 +1,42 @@
+Keep this field blank when using shellhooks.
+. Click the *Additional* tab.
+. Enter the HTTP Content Type.
+  For example, `application/json`, `application/xml` or `text/plain` on the payload you define. The application will not try to convert the content to match the specified content type.
```
Please move the second sentence onto a new line.

melcorr commented on guides/common/modules/ref_shellhooks.adoc:
```diff
@@ -0,0 +1,16 @@
+[id="shellhooks_{context}"]
+= Shellhooks usage details
+
+You can install a {SmartProxy} Shellhooks that exposes executables using a REST HTTPS API.
```
Reads strange. Should it be:
```suggestion
You can install {SmartProxy} Shellhooks that expose executables using a REST HTTPS API.
```

melcorr commented on guides/common/modules/ref_shellhooks.adoc:
```diff
@@ -0,0 +1,16 @@
+
+The HTTPS payload is passed using standard input, optional command line arguments can be provided using X-Shellhook-Arg-1 _to_ N.
+
+The HTTP method must be POST. An example URL would be: `https://{smartproxy-example-com}:{smartproxy_port}/shellhooks/my_script`.
```
Please move second sentence onto a new line.

melcorr commented on guides/common/modules/ref_webhooks-available-events.adoc:
```diff
@@ -0,0 +1,45 @@
+[id="webhooks-available-events_{context}"]
+= Available Events
+
+The following are the webhook events available from the {ProjectWebUI}. `Action` events will trigger webhooks only on `success`, so if an action failed, a webhook is not triggered.
```
Please move the second sentence onto a newline.

melcorr replied:
```suggestion
The following table contains a list of webhook events that are available from the {ProjectWebUI}. `Action` events trigger webhooks only on `success`, so if an action failed, a webhook is not triggered.
```

IBM style says to avoid the future tense.

melcorr commented on guides/common/modules/ref_migrating-from-foreman_webhooks.adoc:
```diff
@@ -0,0 +1,12 @@
+
+The number of events triggered by webhooks is substantially less than with `foreman_hooks`.
+
+The webhooks plugin is designed to be safe, hooks are processed asynchronously so there is minimal risk of tampering with internals of the system. It is not possible to migrate from `foreman_hooks` without creating payloads for each individual webhook script.
```
```suggestion
{Project} webhooks are designed to be safe. 
Webhooks are processed asynchronously so there is minimal risk of tampering with internals of the system. 
It is not possible to migrate from `foreman_hooks` without creating payloads for each individual webhook script.
```

A colon or a full stop is better than the comma here, imo. 
And I moved onto a newline if using comma.

melcorr commented on guides/common/modules/ref_migrating-from-foreman_webhooks.adoc:
```diff
@@ -0,0 +1,12 @@
+
+The webhooks plugin is designed to be safe, hooks are processed asynchronously so there is minimal risk of tampering with internals of the system. It is not possible to migrate from `foreman_hooks` without creating payloads for each individual webhook script.
+
+The webhooks plugin, however, comes with several example payload templates, and together with shellhooks plugin migration should be straightforward. Both script and payload templates must be customized to achieve similar results.
```
This sentence doesn't entirely make sense, especially around this bit
```
together with shellhooks plugin migration should be straightforward. Both script and payload templates must be customized to achieve similar results.
```

I thought the migration that we are talking about here is from Shellhooks to webhooks. 
Please implement the one-sentence-per line rule also.

melcorr reviewed (changes requested):
Hey @neal-timpe, 
Apologies. 
In the end, this was my bad. 
To explain: I tried a few times yesterday and was convinced something was wrong: I believed that items in an earlier commit from you were overwritten. 
This morning, when I looked again, I saw that I was just thrown by a few items that I was convinced you had changed in your commits that were in fact not updated. I've flagged them. 
Elise wrote to me yesterday that you have limited availability to continue working on this. 
Not merging things that are incomplete is a standard practice in both docs and code, so I'm quite bewildered as to how to proceed. 

Traditionally, when we have worked on feature docs, we have often gone through a few reviews because as the information is clarified, some other holes are revealed. 
For you, I've made a number of comments for additional small changes, but as you are aware the shellhooks issue still exists. 
I think that one thing that would help would be to move all the conceptual topics above the procedure for installing webhooks in the chapter.
 
I'm hoping that in Lukas's absence that @ares or @ofedoren might add some insight here that might then clear up the reasons for having webhooks and shellhooks referenced throughout so that we can better explain it to users. I think we are missing vital context here. 

We have a section talking about migration from shellhooks to webhooks. 
We have nothing about the advantages of working with combined webhooks and shellhooks yet the installation procedures are combined. 
Do we want users to blend shellhooks and webhooks?
Should they be treated separately ? 
Do we think users should install both simultaneously?
Should we be encouraging users to move entirely to webhooks and therefore talk about migration but move away from shellhooks (aka delete references to installing them)?

ofedoren commented:
@melcorr, I'll try to answer those questions:
 
 > We have a section talking about migration from shellhooks to webhooks.

That's quite wrong. We don't migrate from `Shellhooks` to `Webhooks`. We migrate from `Hooks` (`foreman_hooks` plugin) to `Webhooks` (`foreman_webhooks` plugin), but with additional `Shellhooks` (`smart_proxy_shellhooks` plugin) to achieve or to preserve almost compatible functionallity which we're dropping (`Hooks`).
> We have nothing about the advantages of working with combined webhooks and shellhooks yet the installation procedures are combined.

It's worth to mention that `Webhooks` is the new way for dealing with events that occur on the server (Satellite/Foreman). It is not compatible with the current way (`Hooks`). To make migration easier we provide additional plugin for proxy (Capsule/Smart Proxy) called `Shellhooks`, so with a combination of these two we can support `Hooks` behaviour on a certain level.

> Do we want users to blend shellhooks and webhooks?

Well, depends on users' use-cases. `Hooks` allows to kind of subscribe on any event that can occur on the server to run scripts on a machine. `Webhooks` allows to subscribe only on a subset of events we provide and support to make HTTP calls to a third party applications (which means any application that accepts HTTP calls). To be able to run scripts on a machine you would simply need to install `Shellhooks`.

> Should they be treated separately ?

`Webhooks` can be installed and used separately, but I'm not so sure about `Shellhooks`. I guess it doesn't make sense to install it without `Webhooks`.

> Do we think users should install both simultaneously?

I guess yes, only if they want similar functionality as with `Hooks`. Although they can be installed and configured separately.

> Should we be encouraging users to move entirely to webhooks and therefore talk about migration but move away from shellhooks (aka delete references to installing them)?

We should encourage users to move to `Webhooks` + `Shellhooks` instead of using `Hooks`.

neal-timpe commented:
@melcorr I've done my best to incorporate the feedback that you gave. Thanks for taking the time to look through this again. I'm just filling in and I'm not on the Satellite team, so I don't know the answers to your questions. I'm not sure if I need to make more changes based on what @ofedoren said. 

I agree and I don't want you to merge things you feel aren't ready. However, my role here is to make changes to someone else's content. I'm happy to make more changes to this if you feel they are necessary.

I also don't want to say I can't, because finding the answers to these questions is possible, but it requires me joining the team and doing a bunch more research.

melcorr commented:
Hey @neal-timpe 
I appreciate you doing what you can. 
Firstly, no one communicated with me what was within or outside of your remit to complete. This PR just appeared alongside another identical PR and I had to then re-do a review to ensure all feedback was carried over.
I'm not on the docs team and have provided thorough reviews both for you and Ian so that these docs are up to the standard expected by Red Hat and in line with the rest of the Satellite docset.
At no point have I been informed that there was no one actually to fulfill the feedback. 
I've also pulled engineers into this discussion to further help with developing the information and they gave up their time only now for their feedback to not make it in. 
The whole idea of having a tech writer develop documentation that it is more than the engineers notes and gives adequate context.
If I were to merge the documentation as is, with the palpable confusion that I flagged and Oleh has clarified, it would give the impression that I stand over incomplete work. 
This is a suboptimal position to be in three months after the initial request to have a tech writer develop the original engineering notes. This has also been a very expensive exercise, if you think of how many people have contributed. 
I appreciate that it isn't necessarily your responsibility, but I don't know what else to do or say. 
Again, thank you for everything you've done. It would be great to have someone with your experience working with us. 
All the best, 
Melanie

maximiliankolb commented on guides/common/modules/proc_creating-a-webhook-template.adoc:
```diff
@@ -0,0 +1,19 @@
++
+A webhook HTTP payload must be created using {Project} template syntax.
++
+For more information, see {ManagingHostsDocURL}appe-red_hat_satellite-managing_hosts-template_writing_reference[Template Writing Reference] in _Managing Hosts_ and for available template macros and methods, visit `/templates_doc` on {ProjectServer}.
```
The anchor has been changed, see `guides/doc-Managing_Hosts/topics/Template_Writing.adoc:4:[[appe-Managing_Hosts-Template_Writing_Reference]]`

```suggestion
For more information, see {ManagingHostsDocURL}appe-Managing_Hosts-Template_Writing_Reference[Template Writing Reference] in _Managing Hosts_ and for available template macros and methods, visit `/templates_doc` on {ProjectServer}.
```

maximiliankolb commented on guides/common/modules/proc_creating-a-webhook.adoc:
```diff
@@ -0,0 +1,42 @@
+[id="creating-a-webhook_{context}"]
+= Creating a Webhook
+
+You can customize events, payloads, HTTP authentication, content type and headers through {ProjectWebUI}. 
```
```suggestion
You can customize events, payloads, HTTP authentication, content type, and headers through {ProjectWebUI}. 
```

See IBM Style Guide: 'Commas between items in a series'. Can you confirm this @melcorr ?

maximiliankolb commented on guides/common/modules/proc_creating-a-webhook.adoc:
```diff
@@ -0,0 +1,42 @@
+. Click *Create Webhook*.
+. Click *Subscribe to* to select an event.
+. Enter a name.
+. Enter a target URL. The target URL can be a dynamic URL.
```
```suggestion
. Enter a target URL.
The target URL can be a dynamic URL.
```

maximiliankolb commented on guides/common/modules/proc_creating-a-webhook.adoc:
```diff
@@ -0,0 +1,42 @@
+. Click *Subscribe to* to select an event.
+. Enter a name.
+. Enter a target URL. The target URL can be a dynamic URL.
+  When using the shellhooks plugin, the URL should be in the form of `https://{smartproxy-example-com}:{smartproxy_port}/shellhooks/my_script`.
```
```suggestion
  When using the shellhooks plugin, the URL should be in the form of `\https://{smartproxy-example-com}:{smartproxy_port}/shellhooks/my_script`.
```

Adding the backslash make the link "unclickable" in asciidoc.

maximiliankolb commented on guides/common/modules/proc_creating-a-webhook.adoc:
```diff
@@ -0,0 +1,42 @@
+. Enter an HTTP method.
+. Check the *Enabled* flag if you want to create an active webhook.
+. Click the *Credentials* tab.
+. Optional: Enter the username and password when HTTP authentication is required.
```
```suggestion
. Optional: If HTTP authentication is required, enter the username and password.
```
IBM Style Guide: purpose before action.

maximiliankolb commented on guides/common/modules/proc_creating-a-webhook.adoc:
```diff
@@ -0,0 +1,42 @@
+. Select *Proxy Authorization* when using shellhooks, otherwise clear this box.
+. Click the *Additional* tab.
+. Enter the HTTP Content Type.
+  For example, `application/json`, `application/xml` or `text/plain` on the payload you define. 
```
```suggestion
For example, `application/json`, `application/xml` or `text/plain` on the payload you define. 
```

maximiliankolb commented on guides/common/modules/proc_creating-a-webhook.adoc:
```diff
@@ -0,0 +1,42 @@
+. Click the *Additional* tab.
+. Enter the HTTP Content Type.
+  For example, `application/json`, `application/xml` or `text/plain` on the payload you define. 
+  The application will not try to convert the content to match the specified content type.
```
```suggestion
The application will not try to convert the content to match the specified content type.
```

maximiliankolb commented on guides/common/modules/proc_creating-a-webhook.adoc:
```diff
@@ -0,0 +1,42 @@
+. Select *Verify SSL* if the server certificate should be verified against the system certificate store or {Project} CA.
+. Select *Proxy Authorization* when using shellhooks, otherwise clear this box.
+. Click the *Additional* tab.
+. Enter the HTTP Content Type.
```
```suggestion
. On the *Additional* tab, enter the HTTP Content Type.
```

maximiliankolb commented on guides/common/modules/proc_creating-a-webhook.adoc:
```diff
@@ -0,0 +1,42 @@
+. Optional: Enter the username and password when HTTP authentication is required.
+. Select *Verify SSL* if the server certificate should be verified against the system certificate store or {Project} CA.
+. Select *Proxy Authorization* when using shellhooks, otherwise clear this box.
+. Click the *Additional* tab.
```
```suggestion
```

maximiliankolb commented on guides/common/modules/proc_creating-a-webhook.adoc:
```diff
@@ -0,0 +1,42 @@
+. Enter the HTTP Content Type.
+  For example, `application/json`, `application/xml` or `text/plain` on the payload you define. 
+  The application will not try to convert the content to match the specified content type.
+. Optional: Provide HTTP headers as JSON. ERB is also allowed.
```
```suggestion
. Optional: Provide HTTP headers as JSON.
ERB is also allowed.
```

maximiliankolb commented on guides/common/modules/proc_creating-a-webhook.adoc:
```diff
@@ -0,0 +1,42 @@
+}
+----
+
+Ensure the content renders to a valid JSON. Also, only pass safe fields like database ID, name or labels which do not include new lines or quote characters.
```
```suggestion
Ensure the content renders to a valid JSON.
Also, only pass safe fields like database ID, name or labels which do not include new lines or quote characters.
```

maximiliankolb commented on guides/common/modules/proc_creating-a-webhook.adoc:
```diff
@@ -0,0 +1,42 @@
+
+Ensure the content renders to a valid JSON. Also, only pass safe fields like database ID, name or labels which do not include new lines or quote characters.
+
+When configuring webhooks with endpoints with non-standard HTTP or HTTPS ports, an SELinux port must be assigned, see {InstallingProjectDocURL}configuring-selinux-to-ensure-access-on-custom-ports_satellite[Configuring SELinux to Ensure Access to Satellite on Custom Ports] in _Installing Satellite Server From A Connected Network_.
```
Please use attributes instead of downstream product names.

maximiliankolb commented on guides/common/modules/proc_installing-webhooks-shellhooks-plugin.adoc:
```diff
@@ -0,0 +1,28 @@
+[id="installing-webhooks-shellhooks-plugin_{context}"]
```
Why does this anchor contain `shellhooks`? I thought webhooks are deprecating shellhooks.

maximiliankolb commented on guides/common/modules/ref_migrating-from-foreman_webhooks.adoc:
```diff
@@ -0,0 +1,14 @@
+[id="migrating-webhooks_{context}"]
+= Migration to Foreman Webhooks
+
+The legacy `foreman_hooks` provides full access to model objects that the webhooks plugin does not intentionally provide.
```
```suggestion
The legacy `foreman_hooks` plugin provides full access to model objects that the webhooks plugin does not intentionally provide.
```

maximiliankolb commented on guides/common/modules/ref_migrating-from-foreman_webhooks.adoc:
```diff
@@ -0,0 +1,14 @@
+[id="migrating-webhooks_{context}"]
```
```suggestion
[id="migrating-to-webhooks_{context}"]
```

maximiliankolb commented on guides/common/modules/ref_migrating-from-foreman_webhooks.adoc:
```diff
@@ -0,0 +1,14 @@
+
+The scope of what is available is limited by the safemode and all objects and macros are both subject to API stability promise and are fully documented.
+
+The number of events triggered by webhooks is substantially less than with `foreman_hooks`.
```
```suggestion
The number of events triggered by webhooks is substantially fewer than with `foreman_hooks`.
```

50% sure, but not a native speaker. @melcorr

maximiliankolb commented on guides/common/modules/ref_migrating-from-foreman_webhooks.adoc:
```diff
@@ -0,0 +1,14 @@
+
+The number of events triggered by webhooks is substantially less than with `foreman_hooks`.
+
+{Project} are designed to be safe.
```
Could you please rephrase this? I am unsure what's the purpose of this statement.

maximiliankolb commented on guides/common/modules/ref_migrating-from-foreman_webhooks.adoc:
```diff
@@ -0,0 +1,14 @@
+{Project} are designed to be safe.
+Webhooks are processed asynchronously so there is minimal risk of tampering with internals of the system. 
+It is not possible to migrate from `foreman_hooks` without creating payloads for each individual webhook script.
+However, the webhooks plugin comes with several example payload templates. You can use the example payloads with shellhooks to simplify migration. 
```
```suggestion
However, the webhooks plugin comes with several example payload templates.
You can use the example payloads with shellhooks to simplify migration. 
```

maximiliankolb commented on guides/common/modules/ref_shellhooks.adoc:
```diff
@@ -0,0 +1,17 @@
+[id="shellhooks_{context}"]
+= Shellhooks usage details
```
```suggestion
= Shellhooks Usage Details
```

maximiliankolb commented on guides/common/modules/ref_shellhooks.adoc:
```diff
@@ -0,0 +1,17 @@
+[id="shellhooks_{context}"]
+= Shellhooks usage details
+
+You can install a {SmartProxy} Shellhooks that expose executables using a REST HTTPS API.
```
Do you mean "shellhooks plugin"?

melcorr commented:
Feel free to reopen this at any time!
Thanks
